pkg/webhooks/validation/controller/controller.go

```diff
@@ -0,0 +1,530 @@
+// Copyright 2019 Istio Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package controller implements a k8s controller for managing the lifecycle of a validating webhook.
+package controller
+
+import (
+	"crypto/x509"
+	"encoding/pem"
+	"errors"
+	"fmt"
+	"io/ioutil"
+	"reflect"
+	"time"
+
+	"github.com/hashicorp/go-multierror"
+	kubeApiAdmission "k8s.io/api/admissionregistration/v1beta1"
+	kubeApiApp "k8s.io/api/apps/v1"
+	kubeApiCore "k8s.io/api/core/v1"
+	kubeApiRbac "k8s.io/api/rbac/v1"
+	kubeErrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/api/meta"
+	kubeApiMeta "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	utilruntime "k8s.io/apimachinery/pkg/util/runtime"
+	"k8s.io/client-go/informers"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/client-go/util/workqueue"
+	"k8s.io/kubectl/pkg/scheme"
+
+	"istio.io/istio/pkg/config/labels"
+	"istio.io/pkg/filewatcher"
+	"istio.io/pkg/log"
+)
+
+var scope = log.RegisterScope("validationController", "validation webhook controller", 0)
+
+type Options struct {
+	Client kubernetes.Interface
+
+	// Istio system namespace in which galley and istiod reside.
+	WatchedNamespace string
+
+	// Periodically resync with the kube-apiserver. Set to zero to disable.
+	ResyncPeriod time.Duration
+
+	// File path to the x509 certificate bundle used by the webhook server
+	// and patched into the webhook config.
+	CAPath string
+
+	// Name of the k8s validatingwebhookconfiguration resource. This should
+	// match the name in the config template.
+	WebhookConfigName string
+
+	// File path to the validatingwebhookconfiguration template.
+	WebhookConfigPath string
+
+	// Name of the service running the webhook server.
+	ServiceName string
+
+	// name of the galley deployment in the watched namespace.
+	// When non-empty the controller will defer reconciling config
+	// until the named deployment no longer exists.
+	GalleyDeploymentName string
+
+	// Name of the ClusterRole that the controller should assign
+	// cluster-scoped ownership to. The webhook config will be GC'd
+	// when this ClusterRole is deleted.
+	ClusterRoleName string
+
+	// Enable galley validation controller.
+	Enabled bool
```

Comment:
Why do we need this? If you don't want to enable it then just don't run it? It doesn't seem like this is really used (meaningfully) either

Reply:
Historical artifact. `controller.Options` was plumbed directly to flags and we had an `enabled` flag.  I've removed the `Enabled` here and will rebase PR 2-of-3 accordingly.